misc/uuidd.8.in
.\" -*- nroff -*-
.\" Copyright 2007 by Theodore Ts'o.  All Rights Reserved.
.\" This file may be copied under the terms of the GNU Public License.
.\"
.TH UUIDD 8 "@E2FSPROGS_MONTH@ @E2FSPROGS_YEAR@" "E2fsprogs version @E2FSPROGS_VERSION@"
.SH NAME
uuidd \- UUID generation daemon
.SH SYNOPSIS
.B uuidd
[
.B \-d
]
[
.B \-p
.I pidfile
]
[
.B \-s
.I socketpath
]
[
.B \-T
.I timeout
]

.B uuidd
[
.B \-r
|
.B \-t
]
[
.B \-n
.I number
]
[
.B \-s
.I socketpath
]

.B uuidd \-k
.SH DESCRIPTION
The
.B uuidd
daemon is used by the UUID library to generate
universally unique identifiers (UUIDs), especially time-based UUID's
in a secure and guaranteed-unique fashion, even in the face of large
numbers of threads trying to grab UUID's running on different CPU's.
.SH OPTIONS
.TP
.B \-d
Run
.B uuidd
in debugging mode.  This prevents uuidd from running as a daemon.
.TP
.B \-k
If a currently uuidd daemon is running, kill it.
.TP
.BI \-n " number"
When issuing a test request to a running uuidd, request a bulk response
of
.I number
UUID's.
.TP
.BI \-p  " pidfile"
Specify the pathname where the pid file should be written.  By default,
the pid file is written to /var/run/uuidd.pid.
.TP
.BI \-s " socketpath"
Specify the pathname used for the unix-domain socket used by uuidd.  By
qdefault, the pathname used is /var/run/uuidd.sock.  This is primarily
for debugging purposes, since the pathname is hard-coded in the libuuid
library.
.TP
.B \-r
Test uuidd by trying to connect to a running uuidd daemon and
request it to return a random-based UUID.
.TP
.B \-t
Test uuidd by trying to connect to a running uuidd daemon and
request it to return a time-based UUID.
.TP
.BI \-T " timeout"
Specify a timeout for uuidd.  If specified, then uuidd will exit after
.I timeout
seconds of inactivity.
.SH AUTHOR
The
.B uuidd
daemon  was written by Theodore Ts'o <tytso@mit.edu>.
.SH AVAILABILITY
.B uuidd
is part of libuuid from the e2fsprogs package and is available from
http://e2fsprogs.sourceforge.net.
.SH "SEE ALSO"
.BR libuuid (3),
.BR uuidgen (1)


ChangeLog
2006-06-24  Theodore Ts'o  <tytso@mit.edu>

	* Release of E2fsprogs 1.40

2007-06-23  Theodore Tso  <tytso@mit.edu>

	* COPYING: Clarify copyright licenses used by the various
		libraries in e2fsprogs.

2007-05-22  Theodore Tso  <tytso@mit.edu>

	* Makefile.in (rpm): Add production rule for e2fsprogs.spec, in
		case it gets deleted.

2007-04-04  Theodore Tso  <tytso@mit.edu>

	* configure, configure.in: Add checks for mmap(), utime(), and
		strdup(), as well as sys/mman.h and utime.h.

2006-11-14  Theodore Tso  <tytso@mit.edu>

	* MCONFIG.in (WFLAGS): Add -fstrict-aliasing and
		-Wno-variadic-macros.  Also use -std=c99 and
		-D_XOPEN_SOURCE=600 and -D_GNU_SOURCE to suppress
		spurious warnings.

2006-11-12  Theodore Tso  <tytso@mit.edu>

	* configure, configure.in (E2FSPROGS_PKGVER): Define
		E2FSPROGS_PKGVER so e2fsprogs.spec can have an appropriate
		version number for WIP releases.

	* e2fsprogs.spec.in: Update/fix spec file so it will work on RHEL 4.

2006-10-22  Theodore Tso  <tytso@mit.edu>

	* configure, configure.in: Add test for jrand48()

	* MCONFIG.in: Add datarootdir definition for compatibility with
		autoconf 2.60.

2006-10-21  Theodore Tso  <tytso@mit.edu>

	* configure, configure.in: Fix -lpthread ordering so that it
		doesn't break on amd64.  (Addresses Debian bug: #393680)

2006-10-01  Theodore Tso  <tytso@mit.edu>

	* configure, configure.in: Include -lpthread if building with
		devmapper support, since the latest devmapper library uses
		-lpthread.  (Addresses Debian bug: #388718)

	* MCONFIG.in: Define DEPLIBBLKID, DEPSTATIC_LIBBLKID, and 
		DEPPROFILED_LIBBLKID.

2006-09-12  Theodore Tso  <tytso@mit.edu>

	* e2fsprogs.spec.in (Group): Use License tag instead of Copyright,
		and specify GPLv2

2006-09-11  Theodore Tso  <tytso@mit.edu>

	* e2fsprogs.spec.in: Remove duplicate entry in the .spec file

2006-08-06  Theodore Tso  <tytso@mit.edu>

	* configure.in: Add test to see if diff supports the -u option

2006-04-09  Theodore Ts'o  <tytso@mit.edu>

	* config/config.guess, config/config.sub: Update to newer versions
		from FSF (2006-02-23)

	* configure.in, Makefile.in, config/parse-types.sh: Make the
		ext2_types.h.in and blkid_types.h.in from config.sys,
		instead of using the subst program.  Create the
		asm_types.h file from the system's
		/usr/include/asm/types.h header file.

2006-03-23  Theodore Ts'o  <tytso@mit.edu>

	* MCONFIG.in (mandir): Define $(man5dir)

2005-10-26  Theodore Ts'o  <tytso@mit.edu>

	* MCONFIG.in, configure.in: Use BUILD_CCFLAGS and BUILD_LDFLAGS
		instead of CCFLAGS and LDFLAGS when to avoid potential
		problems when cross-compiling some of the options are not
		recognized as valid by the host compiler.  (Addresses
		Sourceforge Bug #1261547)

2006-06-30  Theodore Ts'o  <tytso@mit.edu>

	* Release of E2fsprogs 1.38

2005-03-31  Theodore Ts'o  <tytso@mit.edu>

	* configure.in: Add tests for __secure_getenv(), prctl(),
		and sys/prctl.h

2005-03-21  Theodore Ts'o  <tytso@mit.edu>

	* Release of E2fsprogs 1.37

2006-02-05  Theodore Ts'o  <tytso@mit.edu>

	* Release of E2fsprogs 1.36

2005-02-05  Theodore Ts'o  <tytso@mit.edu>

	* configure.in: Make util/gen-tarball executable after it is created.

2005-02-04  Theodore Ts'o  <tytso@mit.edu>

	* configure.in: Remove support for --enable-clear-htree; this was
		only needed during the early development of the htree patch.

	* configure.in, MCONFIG.in: Add support for --enable-maintainer-mode;
		only rebuild configure from configure.in if it is enabled.

2005-02-03  Theodore Ts'o  <tytso@mit.edu>

	* configure.in: Remove support for the (very old) sparc old-bitops
		configure option.

2005-01-19  Matthias Andree <matthias.andree@gmx.de>

	* configure.in: Clean up checks for dirent.d_reclen, ssize_t,
		llseek, lseek64, sockaddr.sa_len and make the lseek checks
		immune to compiler warnings, so that they can be compiled
		with Intel C++ 8.1.

2005-01-18  Theodore Ts'o  <tytso@mit.edu>

	* configure.in: Make configure.in script check for prerequisite
		headers when checking for the presence of sys/disk.h,
		sys/mount.h, and net/if.h, to improve the configure script
		on Solaris and *BSD systems. 

2005-01-17  Theodore Ts'o  <tytso@mit.edu>

	* configure.in: Use AC_PROG_AWK instead of AC_PATH_PROG so that we
		use nawk in preference to awk for Solaris systems.  Use
		AC_PROG_EGREP because Solaris doesn't support "grep -E".

2005-01-09  Theodore Ts'o  <tytso@mit.edu>

	* configure.in: Use AC_CHECK_TYPES instead of the autoconf 2.13
		"broken by design" AC_CHECK_TYPE to look for intptr_t.

2004-12-14  Theodore Ts'o  <tytso@mit.edu>

	* Makefile.in: Add install-strip and install-shlibs-strip targets

	* MCONFIG.in: Add configure-defined variables for MKINSTALLDIRS
		and INSTALL_SCRIPT.  Filter out comments inserted by newer
		versions of gcc when using -M in make depend.

2004-11-30  Theodore Ts'o  <tytso@mit.edu>

	* Makefile.in: Delete autom4te.cache, e2fsprogs.spec,
		ext2ed/Makefile, and po/stamp-po on make distclean

	* Makefile.in: Use Linux-kernel-style makefile output to make it
		easier to see errors/warnings.

	* MCONFIG.in: Add definition for ARGEN

2004-09-18  Theodore Ts'o  <tytso@mit.edu>

	* configure.in: Add --disable-e2initrd-helper flag to control
		whether or not e2initrd_helper program should be
		built/installed.

2004-09-17  Theodore Ts'o  <tytso@mit.edu>

	* Makefile.in: Remove XSI:isms for greater portability.
		(Addresses Debian Bug #255589)

	* config.guess, config.sub: Update to newer version from the FSF
		(2004-06-11)

	* configure.in, configure: Add test for the sys/queue.h header file.

2004-05-04  Theodore Ts'o  <tytso@mit.edu>

	* e2fsprogs.spec.in: Fix up e2fsprogs.spec file so it should work
		on a RH 9.0 system.  Based off of changes proposed by
		Andreas Dilger.

	* configure.in, configure: Enable the fsck wrapper for the Debian
		FreeBSD kernel / GNU userspace port.  (Addresses Debian
		Bug #246738)

	* config.guess, config.sub: Update to newer version from the FSF

misc/uuidd.c
/*
 * uuidd.c --- UUID-generation daemon
 *
 * Copyright (C) 2007  Theodore Ts'o
 *
 * %Begin-Header%
 * This file may be redistributed under the terms of the GNU Public
 * License.
 * %End-Header%
 */

#include <stdio.h>
#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif
#include <unistd.h>
#include <inttypes.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <fcntl.h>
#include <signal.h>
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#else
extern int getopt(int argc, char * const argv[], const char *optstring);
extern char *optarg;
extern int optind;
#endif
#include "uuid/uuid.h"
#include "uuid/uuidd.h"
#include "nls-enable.h"

#ifdef __GNUC__
#define CODE_ATTR(x) __attribute__(x)
#else
#define CODE_ATTR(x)
#endif

static void usage(const char *progname)
{
	fprintf(stderr, _("Usage: %s [-d] [-p pidfile] [-s socketpath] "
			  "[-T timeout]\n"), progname);
	fprintf(stderr, _("       %s [-r|t] [-n num] [-s socketpath]\n"),
		progname);
	fprintf(stderr, _("       %s -k\n"), progname);
	exit(1);
}

static void create_daemon(const char *pidfile_path)
{
	pid_t pid;
	uid_t euid;
	FILE *f;

	pid = fork();
	if (pid == -1) {
		perror("fork");
		exit(1);
	} else if (pid != 0) {
	    exit(0);
	}

	close(0);
	close(1);
	close(2);
	open("/dev/null", O_RDWR);
	open("/dev/null", O_RDWR);
	open("/dev/null", O_RDWR);

	chdir("/");
	(void) setsid();
	euid = geteuid();
	(void) setreuid(euid, euid);

	f = fopen(pidfile_path, "w");
	if (f) {
		fprintf(f, "%d\n", getpid());
		fclose(f);
	}
}

static int read_all(int fd, char *buf, size_t count)
{
	ssize_t ret;
	int c = 0;

	memset(buf, 0, count);
	while (count > 0) {
		ret = read(fd, buf, count);
		if (ret < 0) {
			if ((errno == EAGAIN) || (errno == EINTR))
				continue;
			return -1;
		}
		count -= ret;
		buf += ret;
		c += ret;
	}
	return c;
}

static const char *cleanup_pidfile, *cleanup_socket;

static void terminate_intr(int signo CODE_ATTR((unused)))
{
	(void) unlink(cleanup_pidfile);
	(void) unlink(cleanup_socket);
	exit(0);
}

static void server_loop(const char *socket_path, int debug,
			const char *pidfile_path,
			int timeout, int quiet)
{
	struct sockaddr_un	my_addr, from_addr;
	unsigned char		reply_buf[1024], *cp;
	socklen_t		fromlen;
	int32_t			reply_len = 0;
	uuid_t			uu;
	mode_t			save_umask;
	char			op, str[37];
	int			i, s, ns, len, num;

	if ((s = socket(AF_UNIX, SOCK_STREAM, 0)) < 0) {
		if (!quiet)
			fprintf(stderr, _("Couldn't create unix stream "
					  "socket: %s"), strerror(errno));
		exit(1);
	}

	/*
	 * Create the address we will be binding to.
	 */
	my_addr.sun_family = AF_UNIX;
	strcpy(my_addr.sun_path, socket_path);
	(void) unlink(socket_path);
	save_umask = umask(0);
	if (bind(s, (const struct sockaddr *) &my_addr,
		 sizeof(struct sockaddr_un)) < 0) {
		if (!quiet)
			fprintf(stderr,
				_("Couldn't bind unix socket %s: %s\n"),
				socket_path, strerror(errno));
		exit(1);
	}
	(void) umask(save_umask);

	if (listen(s, 5) < 0) {
		if (!quiet)
			fprintf(stderr, _("Couldn't listen on unix "
					  "socket %s: %s\n"), socket_path,
				strerror(errno));
		exit(1);
	}

	if (!debug) {
		create_daemon(pidfile_path);
		cleanup_pidfile = pidfile_path;
		cleanup_socket = socket_path;
		signal(SIGHUP, terminate_intr);
		signal(SIGINT, terminate_intr);
		signal(SIGPIPE, terminate_intr);
		signal(SIGTERM, terminate_intr);
		signal(SIGALRM, terminate_intr);
	}
	signal(SIGPIPE, SIG_IGN);

	while (1) {
		fromlen = sizeof(from_addr);
		if (timeout > 0)
			alarm(timeout);
		ns = accept(s, (struct sockaddr *) &from_addr, &fromlen);
		alarm(0);
		if (ns < 0) {
			if ((errno == EAGAIN) || (errno == EINTR))
				continue;
			perror("accept");
			exit(1);
		}
		len = read(ns, &op, 1);
		if (len != 1) {
			if (len < 0)
				perror("read");
			else
				printf(_("Error reading from client, "
					 "len = %d\n"), len);
			goto shutdown_socket;
		}
		if ((op == 4) || (op == 5)) {
			if (read_all(ns, (char *) &num, sizeof(num)) != 4)
				goto shutdown_socket;
			if (debug)
				printf(_("operation %d, incoming num = %d\n"),
				       op, num);
		} else if (debug)
			printf("operation %d\n", op);

		switch(op) {
		case UUIDD_OP_GETPID:
			sprintf((char *) reply_buf, "%d", getpid());
			reply_len = strlen((char *) reply_buf)+1;
			break;
		case UUIDD_OP_GET_MAXOP:
			sprintf((char *) reply_buf, "%d", UUIDD_MAX_OP);
			reply_len = strlen((char *) reply_buf)+1;
			break;
		case UUIDD_OP_TIME_UUID:
			num = 1;
			uuid__generate_time(uu, &num);
			if (debug) {
				uuid_unparse(uu, str);
				printf(_("Generated time UUID: %s\n"), str);
			}
			memcpy(reply_buf, uu, sizeof(uu));
			reply_len = sizeof(uu);
			break;
		case UUIDD_OP_RANDOM_UUID:
			num = 1;
			uuid__generate_random(uu, &num);
			if (debug) {
				uuid_unparse(uu, str);
				printf(_("Generated random UUID: %s\n"), str);
			}
			memcpy(reply_buf, uu, sizeof(uu));
			reply_len = sizeof(uu);
			break;
		case UUIDD_OP_BULK_TIME_UUID:
			uuid__generate_time(uu, &num);
			if (debug) {
				uuid_unparse(uu, str);
				printf(_("Generated time UUID %s and %d "
					 "following\n"), str, num);
			}
			memcpy(reply_buf, uu, sizeof(uu));
			reply_len = sizeof(uu);
			memcpy(reply_buf+reply_len, &num, sizeof(num));
			reply_len += sizeof(num);
			break;
		case UUIDD_OP_BULK_RANDOM_UUID:
			if (num < 0)
				num = 1;
			if (num > 1000)
				num = 1000;
			if (num*16 > (int) (sizeof(reply_buf)-sizeof(num)))
				num = (sizeof(reply_buf)-sizeof(num)) / 16;
			uuid__generate_random(reply_buf+sizeof(num), &num);
			if (debug) {
				printf(_("Generated %d UUID's:\n"), num);
				for (i=0, cp=reply_buf+sizeof(num);
				     i < num; i++, cp+=16) {
					uuid_unparse(cp, str);
					printf("\t%s\n", str);
				}
			}
			reply_len = (num*16) + sizeof(num);
			memcpy(reply_buf, &num, sizeof(num));
			break;
		default:
			if (debug)
				printf(_("Invalid operation %d\n"), op);
			goto shutdown_socket;
		}
		write(ns, &reply_len, sizeof(reply_len));
		write(ns, reply_buf, reply_len);
	shutdown_socket:
		close(ns);
	}
}

static int call_daemon(const char *socket_path, int op, unsigned char *buf,
		       int buflen, int *num, const char **err_context)
{
	char op_buf[8];
	int op_len;
	int s;
	ssize_t ret;
	int32_t reply_len = 0;
	struct sockaddr_un srv_addr;

	if (((op == 4) || (op == 5)) && !num) {
		if (err_context)
			*err_context = _("bad arguments");
		errno = EINVAL;
		return -1;
	}

	if ((s = socket(AF_UNIX, SOCK_STREAM, 0)) < 0) {
		if (err_context)
			*err_context = _("socket");
		return -1;
	}

	srv_addr.sun_family = AF_UNIX;
	strcpy(srv_addr.sun_path, socket_path);

	if (connect(s, (const struct sockaddr *) &srv_addr,
		    sizeof(struct sockaddr_un)) < 0) {
		if (err_context)
			*err_context = _("connect");
		close(s);
		return -1;
	}

	if (op == 5) {
		if ((*num)*16 > buflen-4)
			*num = (buflen-4) / 16;
	}
	op_buf[0] = op;
	op_len = 1;
	if ((op == 4) || (op == 5)) {
		memcpy(op_buf+1, num, sizeof(int));
		op_len += sizeof(int);
	}

	ret = write(s, op_buf, op_len);
	if (ret < op_len) {
		if (err_context)
			*err_context = _("write");
		close(s);
		return -1;
	}

	ret = read_all(s, (char *) &reply_len, sizeof(reply_len));
	if (ret < 0) {
		if (err_context)
			*err_context = _("read count");
		close(s);
		return -1;
	}
	if (reply_len < 0 || reply_len > buflen) {
		if (err_context)
			*err_context = _("bad response length");
		close(s);
		return -1;
	}
	ret = read_all(s, (char *) buf, reply_len);

	if ((ret > 0) && (op == 4)) {
		if (reply_len >= (int) (16+sizeof(int)))
			memcpy(buf+16, num, sizeof(int));
		else
			*num = -1;
	}
	if ((ret > 0) && (op == 5)) {
		if (*num >= (int) sizeof(int))
			memcpy(buf, num, sizeof(int));
		else
			*num = -1;
	}

	close(s);

	return ret;
}


int main(int argc, char **argv)
{
	const char	*socket_path = UUIDD_SOCKET_PATH;
	const char	*pidfile_path = UUIDD_PIDFILE_PATH;
	const char	*err_context;
	unsigned char	buf[1024], *cp;
	char   		str[37], *tmp;
	uuid_t		uu;
	uid_t		uid;
	gid_t 		gid;
	int		i, c, ret;
	int		debug = 0, do_type = 0, do_kill = 0, num = 0;
	int		timeout = 0, quiet = 0, drop_privs = 0;

#ifdef ENABLE_NLS
	setlocale(LC_MESSAGES, "");
	setlocale(LC_CTYPE, "");
	bindtextdomain(NLS_CAT_NAME, LOCALEDIR);
	textdomain(NLS_CAT_NAME);
#endif

	while ((c = getopt (argc, argv, "dkn:qp:s:tT:r")) != EOF) {
		switch (c) {
		case 'd':
			debug++;
			drop_privs++;
			break;
		case 'k':
			do_kill++;
			drop_privs++;
			break;
		case 'n':
			num = strtol(optarg, &tmp, 0);
			if ((num < 0) || *tmp) {
				fprintf(stderr, _("Bad number: %s\n"), optarg);
				exit(1);
			}
		case 'p':
			pidfile_path = optarg;
			drop_privs++;
			break;
		case 'q':
			quiet++;
			break;
		case 's':
			socket_path = optarg;
			drop_privs++;
			break;
		case 't':
			do_type = UUIDD_OP_TIME_UUID;
			drop_privs++;
			break;
		case 'T':
			timeout = strtol(optarg, &tmp, 0);
			if ((timeout < 0) || *tmp) {
				fprintf(stderr, _("Bad number: %s\n"), optarg);
				exit(1);
			}
			break;
		case 'r':
			do_type = UUIDD_OP_RANDOM_UUID;
			drop_privs++;
			break;
		default:
			usage(argv[0]);
		}
	}
	uid = getuid();
	if (uid && drop_privs) {
		gid = getgid();
#ifdef HAVE_SETRESUID
		setresuid(uid, uid, uid);
#else
		setreuid(uid, uid);
#endif
#ifdef HAVE_SETRESGID
		setresgid(gid, gid, gid);
#else
		setregid(gid, gid);
#endif
	}
	if (num && do_type) {
		ret = call_daemon(socket_path, do_type+2, buf,
				  sizeof(buf), &num, &err_context);
		if (ret < 0) {
			printf(_("Error calling uuidd daemon (%s): %s\n"),
			       err_context, strerror(errno));
			exit(1);
		}
		if (do_type == UUIDD_OP_TIME_UUID) {
			if (ret != sizeof(uu) + sizeof(num))
				goto unexpected_size;

			uuid_unparse(buf, str);

			printf(_("%s and subsequent %d UUID's\n"), str, num);
		} else {
			printf(_("List of UUID's:\n"));
			cp = buf + 4;
			if (ret != sizeof(num) + num*sizeof(uu))
				goto unexpected_size;
			for (i=0; i < num; i++, cp+=16) {
				uuid_unparse(cp, str);
				printf("\t%s\n", str);
			}
		}
		exit(0);
	}
	if (do_type) {
		ret = call_daemon(socket_path, do_type, (unsigned char *) &uu,
				  sizeof(uu), 0, &err_context);
		if (ret < 0) {
			printf(_("Error calling uuidd daemon (%s): %s\n"),
			       err_context, strerror(errno));
			exit(1);
		}
		if (ret != sizeof(uu)) {
		unexpected_size:
			printf(_("Unexpected reply length from server %d\n"),
			       ret);
			exit(1);
		}
		uuid_unparse(uu, str);

		printf("%s\n", str);
		exit(0);
	}

	/*
	 * Check to make sure there isn't another daemon running already
	 */
	ret = call_daemon(socket_path, 0, buf, sizeof(buf), 0, 0);
	if (ret > 0) {
		if (do_kill && ((do_kill = atoi((char *) buf)) > 0)) {
			ret = kill(do_kill, SIGTERM);
			if (ret < 0) {
				if (!quiet)
					fprintf(stderr,
						_("Couldn't kill uuidd running "
						  "at pid %d: %s\n"), do_kill,
						strerror(errno));
				exit(1);
			}
			if (!quiet)
				printf(_("Killed uuidd running at pid %d\n"),
				       do_kill);
			exit(0);
		}
		if (!quiet)
			printf(_("uuidd daemon already running at pid %s\n"),
			       buf);
		exit(1);
	}
	if (do_kill)
		exit(0);	/* Nothing to kill */

	server_loop(socket_path, debug, pidfile_path, timeout, quiet);
	return 0;
}
